### ISSUE TITLE
Wifidog-ng redirect loop

### ISSUE THREAD

**Issue by hugorut** [zhaojh329/wifidog-ng#14]
Firmware: `OpenWrt 18.06.0-rc2, r7141-e4d0ee5af5`
Wifidog-ng Version: `kmod-wifidog-ng - 4.9.111+1.5.6-1 wifidog-ng-openssl - 1.5.6-1`

Hi there I am struggling with an issue with `wifidog-ng`. 

I have setup `wifidog-ng` auth server on a server and all seems to be working correctly. However sometimes the auth server enters a reload loop. Like below:

![](https://j.gifs.com/pQ5jw6.gif)

Below is my `/etc/config/wifidog-ng.conf` file.

![](https://i.imgur.com/arWFpHs.png)

I even tried whitelisting the domain in the config file. Which worked for the whitelisting but did not solve the reload loop issue.

I thought this might be something to do with my server and using variable IPs but I locked them down to one static IP and it still did not work. I also followed more verbose logging on my server and it doesn't actually appear to be hitting the auth application. So something is going down with wifidog. 

I tried to follow the openwrt logs using `logread -f` to see if there was anything that could point me in the right direction however it didn't give me anything useful.

The only way that I can solve this problem is by restarting wifidog `/etc/init.d/wifidog-ng restart` and then navigating using the internet within a short timeframe (< 5mins). This seems to allow the page to load and I can authenticate successfully and then browse.

Any help on how I could track down what is going on would be amazing as I'm currently at a loss. 

Thanks

**Comment by zhaojh329** [comment]
I can't locate the problem from the current information, I think you need to analyze your server to see which step is looping.

**Comment by hugorut** [comment]
@zhaojh329 so have tracked down a little further and it appears that wifidog is inserting a javascript into the login page of the authserver which causes an endless refresh loop. See below:
![](https://i.imgur.com/0EhU1lO.png)
I've located the code snippet that does this to this method: https://github.com/zhaojh329/wifidog-ng/blob/9a8a58ba56d7924396c8b6248f5b0dc3d78fe186/src/auth.c#L122 
Which looks like it inserts this code into the page if it receives a 404 response. Is there anything that my server's response could be doing to execute this method? (The normal response for the /login route is 200).  I'm trying to work it out the problem from source code but I'm no C expert and both `wifidog-ng` and the `libuhttpd` package it relies on are changing heavily so it proves rather tricky to trace down the functionality. Any help would be appreciated.

**Comment by zhaojh329** [comment]
Please confirm that the URL interface is correct.
https://github.com/zhaojh329/wifidog-ng/tree/C#auth

**Comment by hugorut** [comment]
Yes the auth interface is correct. 
```php
$url = "http://".$request->input('gw_address').":".$request->input('gw_port')."/wifidog/auth?token=$token";
```

So if I understand you correctly you believe this is happening because the authserver is calling an incorrect route on the gateway. i.e. 

1.  user attempts to request bbc.co.uk
2. wifidog intercepts and redirects to authserver address
3. authserver calls unknown route on the gateway
4. gateway executes javascript to reload the authserver login address
5. falls back to step 3. 

I'm not sure how this could be happening. However when the redirect loop happens the authserver is not calling the gateway. This happens **after** login. I can't even get to this stage as the authserver won't load. 

I have also tested switching the authserver address to completely different url and the redirect loop still occurs.

**Comment by hugorut** [comment]
@zhaojh329 is there anyway that I can enable a more verbose output from `wifidog-ng` about the inbound connections that it is handling?

**Comment by zhaojh329** [comment]
It is recommended to use the latest Lua version.

**Comment by hugorut** [comment]
ok so I found the issue. The problem was that the dns I was for my authserver was not using a static IP. Instead it was dynamically cycling through a list of servers so the outward facing ips often changed. This also the case with other servers that I tested. i.e. bbc, s3 hosted sites e.t.c. 

`wifidog-ng` on boot checks the authserver dns and adds the know ips at that time to the *allow* `wifidog-ng-ip` ipset. However the later down the line when those ips become invalid wifidog will enter a never-ending loop as described below.

1. new terminal requests website x
2. wifidog-ng checks that x is not within the whitelisted ipset
3. it reloads the page using js to point to the authserver
4. the page then requests the dns **which now has a new ip address to the whitelisted**
5. wifidog-ng executes the page reload again

The solution for me was to put my authserver behind a proxy which served static ips for the inbound traffic. 

However there might be something that can be done within the `wifidog-ng` internals to periodically update the ipset whitelist of the authserver as lots of sites nowadays use dynamic ip ranges. This might also need to be implemented for generic whitelisted ips as well. For example fonts.googleapis.com does not serve a unchanging ip.
